model/service.go

```diff
@@ -21,40 +153,183 @@ import (
+	// really? the scenario we are worried about is a kubernetes loophole,
+	// where a looney user can define two ports for a service such that
+	// both ports have the same port number but a different name This is a
+	// corner case and more like an abuse of kubernetes.
+
+	// FIXME detect this stuff in validation and eliminate such duplicates
+	Name string `json:"name,omitempty"`
+
+	// Port number where the service can be reached Does not necessarily
+	// map to the corresponding port numbers for the instances behind the
+	// service. See endpoint definition below.
+	Port int `json:"port"`
+
+	// Protocol to be used for the port. How the protocol associated with a
+	// port is identified depends heavily on the underlying platform. For
+	// e.g., in Kubernetes, the port definition allows the user to define a
+	// name for the port where the name can indicate various higher level
+	// protocols like Http, Http2, gRPC, etc.
+	Protocol Protocol `json:"protocol,omitempty"`
+}
+
+// PortList is a set of ports
+type PortList []*Port
+
+// Protocol defines network protocols for ports
+type Protocol string
+
+// Protocols used by the services
+const (
+	ProtocolGRPC  Protocol = "GRPC"
+	ProtocolHTTPS Protocol = "HTTPS"
+	ProtocolHTTP2 Protocol = "HTTP2"
+	ProtocolHTTP  Protocol = "HTTP"
+	ProtocolTCP   Protocol = "TCP"
+	ProtocolUDP   Protocol = "UDP"
+)
+
+// Endpoint defines a IP:port associated with an instance of the service. This is the address to which the
+// request/connection from the caller will be routed to. If a service has multiple ports, then the same
+// instance IP is listening on multiple ports (one corresponding to each service port). However,
+// we do not group such ports here. Instead we use a one endpoint per service port.
+
+// For e.g., if catalog.mystore.com is accessible through port 80 and 8080,
+// and it maps to an instance with IP 172.16.0.1, such that connections to
+// port 80 are forwarded to port 55446, and connections to port 8080 are
+// forwarded to port 33333,
+
+// then internally, we have two two endpoint structs for the
+// service catalog.mystore.com
+//  --> 172.16.0.1:54546 (with ServicePort pointing to 80) and
+//  --> 172.16.0.1:33333 (with ServicePort pointing to 8080)
+type Endpoint struct {
+	// Address of the network endpoint, typically an IPv4 address
+	Address string `json:"ip_address,omitempty"`
+
+	// Port number where this instance is listening for connections This
+	// need not be the same as the port where the service is accessed.
+	// e.g., catalog.mystore.com:8080 -> 172.16.0.1:55446
+	Port int `json:"port"`
+
+	// Port declaration from the service declaration This is the port for
+	// the service associated with this instance (e.g.,
+	// catalog.mystore.com)
+	ServicePort *Port `json:"port"`
+}
+
+// Tag is non empty set of key=value pair attributes (e.g., Labels in
+// kubernetes/mesos) that are assigned to pods/VMs etc. Each version of a
+// service can be uniquely identified by a set of tags
```

Comment:
do not describe tag by its implementation.
this is bound to change and shouldn't be fixed. tag is just a tag.